Sources/ArcGISToolkit/Components/Form/FormInputs/DateTimeEntry.swift

```diff
@@ -0,0 +1,208 @@
+    /// Creates a view for a date and time (if applicable) entry.
+    /// - Parameters:
+    ///   - element: The field's parent element.
+    ///   - input: The input configuration of the field.
+    init(element: FieldFeatureFormElement, input: DateTimePickerFeatureFormInput) {
+        self.element = element
+        self.input = input
+    }
+    
+    var body: some View {
+        Group {
+            FormElementHeader(element: element)
+                .padding([.top], elementPadding)
+            
+            dateViewer
+            
+            if isEditing {
+                dateEditor
+            }
+            
+            footer
+        }
+        .padding([.bottom], elementPadding)
+        .onAppear {
+            if let date = model.feature?.attributes[element.fieldName] as? Date {
+                self.date = date
+            }
+        }
+        .onChange(of: date) { newDate in
+            requiredValueMissing = element.required && newDate == nil
+            model.feature?.setAttributeValue(newDate, forKey: element.fieldName)
+        }
+    }
+    
+    /// Controls for modifying the date selection.
+    @ViewBuilder var dateEditor: some View {
+        HStack {
+            todayOrNowButton
+            Spacer()
+            doneButton
+        }
+        datePicker
+            .datePickerStyle(.graphical)
+    }
+    
+    /// Elements for viewing the date selection.
+    /// - Note: Secondary foreground color is used across entry views for consistency.
+    @ViewBuilder var dateViewer: some View {
+        HStack {
+            TextField(
+                element.label,
+                text: Binding { date == nil ? "" : formattedDate } set: { _ in },
+                prompt: .noValue.foregroundColor(.secondary)
+            )
+            .disabled(true)
+            
+            if date == nil {
+                Image(systemName: "calendar")
+                    .foregroundColor(.secondary)
+            } else {
+                ClearButton { date = nil }
+            }
+        }
+        .formTextEntryStyle()
+        .onTapGesture {
+            if date == nil { date = .now }
+            withAnimation { isEditing = true }
+        }
+    }
+    
+    /// Controls for making a specific date selection.
+    @ViewBuilder var datePicker: some View {
+        let components: DatePicker.Components = input.includeTime ? [.date, .hourAndMinute] : [.date]
+        if let min = input.min, let max = input.max {
+            DatePicker(selection: Binding($date)!, in: min...max, displayedComponents: components) { }
+        } else if let min = input.min {
+            DatePicker(selection: Binding($date)!, in: min..., displayedComponents: components) { }
+        } else if let max = input.max {
+            DatePicker(selection: Binding($date)!, in: ...max, displayedComponents: components) { }
+        } else {
+            DatePicker(selection: Binding($date)!, displayedComponents: components) { }
+        }
+    }
```

Comment:
This looks a little cluttered to me, what about putting the range calculation in a different method:
```suggestion
    /// Controls for making a specific date selection.
    @ViewBuilder var datePicker: some View {
        let components: DatePicker.Components = input.includeTime ? [.date, .hourAndMinute] : [.date]
        if let range = dateRange {
            DatePicker(selection: Binding($date)!, in: range, displayedComponents: components) { }
        } else {
            DatePicker(selection: Binding($date)!, displayedComponents: components) { }
        }
    }
    
    var dateRange: ClosedRange<Date>? {
        if let min = input.min, let max = input.max {
            return min...max
        } else if let min = input.min {
            return min...Date.distantFuture
        } else if let max = input.max {
            return Date.distantPast...max
        } else {
            return nil
        }
    }
```

Reply:
I was wondering if the `nil` case above could be `return Date.distantPast...Date.distantFuture`; would that be equivalent? You could get rid of the `if let dateRange...` above.

Reply:
Unless we need to support dates outside this range, I think we can go with that.

<p align=center>
<img width="611" alt="Screenshot 2023-07-21 at 17 49 53" src="https://github.com/Esri/arcgis-maps-sdk-swift-toolkit/assets/16397058/ef36bd9f-04df-4381-b818-fae94041975c">
</p>